### PR TITLE
Implement floating point const and add

### PR DESCRIPTION
**Pull request by chfast**
A symbolic milestone starting floating point support implementation.

### CODE REVIEW

File: lib/fizzy/value.hpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    constexpr Value(int64_t v) noexcept : i64{static_cast<uint64_t>(v)} {}
+    constexpr Value(int32_t v) noexcept : i64{static_cast<uint32_t>(v)} {}
+
+    constexpr Value(float v) noexcept : f32{v} {}
+
+    constexpr operator uint64_t() const noexcept { return i64; }
+
+    /// Get the value as the given type. Handy in templates.
+    template <typename T>
+    constexpr T as() const noexcept
```
Comment (gumb0) [inline lib/fizzy/value.hpp:29]: Can this be left without default implementation and then have specializations for `uint32_t` and `uint64_t` similar to `float`?

**Comment by chfast** [comment]
Yes, makes sense.

File: lib/fizzy/value.hpp (hunk 1 above)
Comment (chfast) [reply lib/fizzy/value.hpp]: Done in #423.

**Comment by axic** [comment]
Doesnt this need spectest expectation update in CI?

**Comment by chfast** [comment]
> Doesnt this need spectest expectation update in CI?

I was expecting that, but no. What is the conditions for skipping a spec test?

File: test/unittests/value_test.cpp [hunk 2]
```diff
@@ -41,6 +41,12 @@ TEST(value, constructor_from_signed_ints)
     EXPECT_EQ(Value{int64_t{-3}}.i64, 0xfffffffffffffffd);
 }
 
+TEST(value, constructor_from_floating_points)
+{
+    EXPECT_EQ(Value{123.456f}.f32, 123.456f);
+    EXPECT_EQ(Value{123.456789001}.f64, 123.456789001);
```
Comment (axic) [inline test/unittests/value_test.cpp:47]: Would be nice to have a test for min/max.

Comment (chfast) [reply test/unittests/value_test.cpp:47]: You mean the total max/min, or the biggest mantissa only?

Comment (chfast) [reply test/unittests/value_test.cpp:47]: I used some special values in tests.

Comment (axic) [reply test/unittests/value_test.cpp:47]: Basically something which can not be represented in f32. `f64::max()` is fine, but not sure what that is actually storing, I believe a large exponent?

File: lib/fizzy/parser_expr.cpp [hunk 3]
```diff
@@ -799,14 +805,27 @@ parser_result<Code> parse_expr(const uint8_t* pos, const uint8_t* end, FuncIdx f
...
         {
             int64_t value;
             std::tie(value, pos) = leb128s_decode<int64_t>(pos, end);
             push(code.immediates, static_cast<uint64_t>(value));
             break;
         }
+        case Instr::f32_const:
+        {
+            uint32_t value;
+            std::tie(value, pos) = read<decltype(value)>(pos, end);
```
Comment (gumb0) [inline lib/fizzy/parser_expr.cpp:818]: what is the point of using `decltype` here if `value` is declared on the previous line?

File: lib/fizzy/parser_expr.cpp [hunk 4]
```diff
@@ -27,6 +27,18 @@ inline void push(bytes& b, T value)
     b.append(storage, sizeof(storage));
 }
 
+template <typename T>
+inline parser_result<T> read(const uint8_t* pos, const uint8_t* end)
```
Comment (gumb0) [inline lib/fizzy/parser_expr.cpp:31]: I would call it maybe `read_value` or `read_const`

File: test/unittests/value_test.cpp [hunk 5]
```diff
@@ -50,6 +56,40 @@ TEST(value, as_integer)
...
+    f = std::numeric_limits<float>::min();
+    EXPECT_EQ(Value{f}.f32, f);
+    EXPECT_EQ(Value{f}.as<float>(), f);
+    f = std::numeric_limits<float>::max();
+    EXPECT_EQ(Value{f}.f32, f);
+    EXPECT_EQ(Value{f}.as<float>(), f);
+    f = std::numeric_limits<float>::denorm_min();
+    EXPECT_EQ(Value{f}.f32, f);
+    EXPECT_EQ(Value{f}.as<float>(), f);
+
```
Comment (gumb0) [inline test/unittests/value_test.cpp:77]: Could also add `numeric_limits<float>::lowest()`

File: lib/fizzy/parser_expr.cpp (hunk 4 above)
Comment (chfast) [reply lib/fizzy/parser_expr.cpp:31]: Or `read_float`.

File: lib/fizzy/parser_expr.cpp (hunk 3 above)
Comment (chfast) [reply lib/fizzy/parser_expr.cpp:818]: Easier to copy-paste to `Instr::f64_const`.

File: test/unittests/value_test.cpp (hunk 5 above)
Comment (chfast) [reply test/unittests/value_test.cpp:80]: Done.

**Pull request merged**